module/VuFind/src/VuFind/Mailer/Factory.php

```diff
@@ -76,8 +76,8 @@ protected function getDSN(array $config): string
         $protocol = ($config['Mail']['secure'] ?? false) ? 'smtps' : 'smtp';
         $dsn = "$protocol://";
         if (
-            ($username = $config['Mail']['username'] ?? null)
-            && ($password = $this->getSecretFromConfig($config['Mail'], 'password'))
+            ($username = rawurlencode($config['Mail']['username']) ?? null)
+            && ($password = rawurlencode($this->getSecretFromConfig($config['Mail'], 'password')))
```

Comment:
```suggestion
            ('' !== ($username = rawurlencode($config['Mail']['username']) ?? ''))
            && ('' !== ($password = rawurlencode($this->getSecretFromConfig($config['Mail'], 'password') ?? '')))
```
rawurlencode expects a string, so you need to use `?? ''` for both username and password (getSecretFromConfig can also return null) to avoid passing a null to it.

As a side note, these checks, including the existing ones, aren't quite right in that they should actually compare username and password against an empty string to allow anything that PHP thinks is "falsy" to work as username or password. That's why the suggestion adds the comparison.

Reply:
Oh, excellent nuance. Thanks, Ere!